lib/vector-core/src/transform/mod.rs

```diff
@@ -382,54 +407,81 @@ impl ByteSizeOf for TransformOutputsBuf {
 }
 
 #[derive(Debug, Default)]
-pub struct OutputBuffer(Vec<Event>);
+pub struct OutputBuffer(Vec<EventArray>);
 
 impl OutputBuffer {
     pub fn with_capacity(capacity: usize) -> Self {
         Self(Vec::with_capacity(capacity))
     }
 
     pub fn push(&mut self, event: Event) {
-        self.0.push(event);
+        // Coalesce multiple pushes of the same type into one array.
+        match (event, self.0.last_mut()) {
+            (Event::Log(log), Some(EventArray::Logs(logs))) => {
+                logs.push(log);
+            }
+            (Event::Metric(metric), Some(EventArray::Metrics(metrics))) => {
+                metrics.push(metric);
+            }
+            (event, _) => {
+                self.0.push(event.into());
+            }
+        }
     }
 
     pub fn append(&mut self, events: &mut Vec<Event>) {
-        self.0.append(events);
+        for event in events.drain(..) {
+            self.push(event);
+        }
     }
 
     pub fn extend(&mut self, events: impl Iterator<Item = Event>) {
-        self.0.extend(events);
+        for event in events {
+            self.push(event);
+        }
     }
 
     pub fn pop(&mut self) -> Option<Event> {
-        self.0.pop()
+        let result = self.0.last_mut().and_then(EventArray::pop);
+        // Make sure to clean up empty arrays
+        if self.0.last().map_or(false, EventArray::is_empty) {
+            self.0.pop();
+        }
```

Comment:
Would like to see a unit test for this, would also help not to regress on this behavior in the future.

Reply:
Done in fa26f3452